# Bài 1: Làm Quen Với PyTorch

**Họ tên:** Huỳnh Nhật Quang
**MSSV:** 3124411245
**Môn học:** Machine Learning
**Trường:** Đại học Sài Gòn (SGU)

## 2. Bài tập thực hành

### Bài 1. Tạo Tensor cơ bản

In [1]:
import torch

a = torch.tensor(
    [[1, 2, 3],
     [4, 5, 6]], dtype=torch.int32
)
print("Tensor a:\n", a)

# YÊU CẦU: Cho biết kích thước, số chiều và kiểu dữ liệu
print("\n--- Yêu cầu Bài 1 ---")
print("Kích thước (Shape):", a.shape)
print("Số chiều (ndim):", a.ndim)
print("Kiểu dữ liệu (dtype):", a.dtype)

Tensor a:
 tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)

--- Yêu cầu Bài 1 ---
Kích thước (Shape): torch.Size([2, 3])
Số chiều (ndim): 2
Kiểu dữ liệu (dtype): torch.int32


### Bài 2. Tạo các Tensor thường dùng

In [2]:
import torch

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)

print("torch.rand:\n", a)
print("torch.randn:\n", b)
print("torch.zeros:\n", c)
print("torch.ones:\n", d)
print("torch.eye:\n", e)

# YÊU CẦU: Tạo thêm Tensor 5x5 chứa toàn giá trị 10
f = torch.full((5, 5), 10)
print("\nTensor 5x5 toàn giá trị 10:\n", f)

torch.rand:
 tensor([[0.3962, 0.9772, 0.5544, 0.4439],
        [0.1852, 0.9896, 0.3192, 0.0508],
        [0.1298, 0.6937, 0.8906, 0.7993]])
torch.randn:
 tensor([[ 0.2249, -0.3736, -0.0264,  0.5242],
        [ 0.4228, -0.6468,  0.3692,  1.1318],
        [ 1.1451, -0.5719, -1.9017, -0.3842]])
torch.zeros:
 tensor([[0., 0., 0.],
        [0., 0., 0.]])
torch.ones:
 tensor([[1., 1., 1.],
        [1., 1., 1.]])
torch.eye:
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

Tensor 5x5 toàn giá trị 10:
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


**Nhận xét (Yêu cầu Bài 2):**
- `torch.rand()`: Sinh ngẫu nhiên các giá trị từ phân phối đều (uniform distribution) trong khoảng [0, 1).
- `torch.randn()`: Sinh ngẫu nhiên các giá trị từ phân phối chuẩn tắc (standard normal distribution) với trung bình = 0 và độ lệch chuẩn = 1.

### Bài 3. Kiểm tra thông tin Tensor

In [3]:
import torch

a = torch.zeros(2, 3, 4)
print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


**Giải thích (Yêu cầu Bài 3):**
Ý nghĩa của `torch.Size([2, 3, 4])` là Tensor này có 3 chiều (3D tensor). Cụ thể, nó chứa 2 ma trận (khối), mỗi ma trận có kích thước 3 hàng và 4 cột.

### Bài 4. Slicing và Indexing

In [4]:
import torch

a = torch.randn(3, 4, 5)
print("Original tensor shape:", a.shape)

print("\nTensor a[1] shape:", a[1].shape)
print("Tensor a[1:, 2:4] shape:", a[1:, 2:4].shape)

# YÊU CẦU: Thử lấy hai phần tử cuối cùng trên chiều thứ ba
# Lấy tất cả các phần tử ở chiều 1 và chiều 2, nhưng chỉ lấy 2 cột cuối ở chiều 3
last_two = a[:, :, -2:]
print("\nHai phần tử cuối cùng trên chiều thứ ba (shape):", last_two.shape)
print(last_two)

Original tensor shape: torch.Size([3, 4, 5])

Tensor a[1] shape: torch.Size([4, 5])
Tensor a[1:, 2:4] shape: torch.Size([2, 2, 5])

Hai phần tử cuối cùng trên chiều thứ ba (shape): torch.Size([3, 4, 2])
tensor([[[-0.4439, -0.7315],
         [-1.5292,  0.6233],
         [-0.3561,  1.6941],
         [-0.1799, -2.4169]],

        [[-0.3697, -0.0606],
         [-0.4937,  0.4025],
         [-0.1805, -0.1225],
         [-1.0230,  1.1382]],

        [[-1.2203,  0.8951],
         [ 0.8750, -0.6157],
         [-0.7165,  0.1302],
         [ 0.6612,  0.2962]]])


### Bài 5. Reshape và Flatten

In [5]:
import torch

a = torch.randn(3, 4)
print("Original:")
print(a)

print("\nFlatten:")
print(a.ravel())

print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

# YÊU CẦU:
print("\n--- Yêu cầu Bài 5 ---")
print("Số phần tử trước khi reshape:", a.numel())
print("Số phần tử sau khi reshape (3,2,2):", a.reshape(3,2,2).numel())

print("\nReshape thành kích thước 2 x 6:")
print(a.reshape(2, 6))

Original:
tensor([[ 0.2356,  0.5255,  0.1527,  0.3477],
        [-0.9046,  0.3642,  0.7881,  0.4403],
        [ 3.6921,  1.1671,  1.0963, -0.0086]])

Flatten:
tensor([ 0.2356,  0.5255,  0.1527,  0.3477, -0.9046,  0.3642,  0.7881,  0.4403,
         3.6921,  1.1671,  1.0963, -0.0086])

Reshape to 3 x 2 x 2:
tensor([[[ 0.2356,  0.5255],
         [ 0.1527,  0.3477]],

        [[-0.9046,  0.3642],
         [ 0.7881,  0.4403]],

        [[ 3.6921,  1.1671],
         [ 1.0963, -0.0086]]])

--- Yêu cầu Bài 5 ---
Số phần tử trước khi reshape: 12
Số phần tử sau khi reshape (3,2,2): 12

Reshape thành kích thước 2 x 6:
tensor([[ 0.2356,  0.5255,  0.1527,  0.3477, -0.9046,  0.3642],
        [ 0.7881,  0.4403,  3.6921,  1.1671,  1.0963, -0.0086]])


### Bài 6. Phép toán và nhân ma trận

In [6]:
import torch

a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:\n", a)
print("b:\n", b)
print("\na + b:\n", a + b)
print("\na ** 2:\n", a ** 2)
print("\na / b:\n", a / b)

print("\nMatrix multiplication (a @ b.T):\n", a @ b.T)
print("Shape của kết quả a @ b.T:", (a @ b.T).shape)

a:
 tensor([[ 0.9000, -0.1813, -1.3419],
        [ 1.1817, -1.4428, -0.1211]])
b:
 tensor([[ 1.4081,  1.1290,  0.4892],
        [-0.0653, -1.6699, -0.4060]])

a + b:
 tensor([[ 2.3080,  0.9477, -0.8527],
        [ 1.1164, -3.1127, -0.5271]])

a ** 2:
 tensor([[0.8100, 0.0329, 1.8007],
        [1.3964, 2.0817, 0.0147]])

a / b:
 tensor([[  0.6392,  -0.1606,  -2.7428],
        [-18.0945,   0.8640,   0.2982]])

Matrix multiplication (a @ b.T):
 tensor([[ 0.4061,  0.7888],
        [-0.0242,  2.3814]])
Shape của kết quả a @ b.T: torch.Size([2, 2])


**Nhận xét (Yêu cầu Bài 6):**
- Phép toán `a * b`: Là phép nhân từng phần tử tương ứng (element-wise multiplication) của hai ma trận cùng kích thước.
- Phép toán `a @ b.T`: Là phép nhân ma trận (matrix multiplication). Kích thước của `a` là (2, 3), `b.T` là (3, 2) nên kết quả có shape là (2, 2).

### Bài 7. Các hàm thống kê trên Tensor

In [7]:
import torch

a = torch.randn(3, 4)
print("Original:\n", a)

print("\nMean (dim=0):")
print(torch.mean(a, dim=0))

print("\nStandard deviation (dim=0):")
print(torch.std(a, dim=0))

print("\nCumulative sum (dim=0):")
print(torch.cumsum(a, dim=0))

# YÊU CẦU: Thay dim=0 bằng dim=1
print("\n--- Yêu cầu Bài 7 ---")
print("Mean (dim=1):\n", torch.mean(a, dim=1))
print("Standard deviation (dim=1):\n", torch.std(a, dim=1))
print("Cumulative sum (dim=1):\n", torch.cumsum(a, dim=1))

Original:
 tensor([[ 0.3086, -0.0070,  0.8604,  0.4072],
        [ 0.4686,  1.1621, -0.3260, -0.1108],
        [ 1.6072,  0.2191, -0.8565, -1.0404]])

Mean (dim=0):
tensor([ 0.7948,  0.4581, -0.1074, -0.2480])

Standard deviation (dim=0):
tensor([0.7081, 0.6201, 0.8791, 0.7334])

Cumulative sum (dim=0):
tensor([[ 0.3086, -0.0070,  0.8604,  0.4072],
        [ 0.7772,  1.1551,  0.5344,  0.2963],
        [ 2.3844,  1.3742, -0.3221, -0.7440]])

--- Yêu cầu Bài 7 ---
Mean (dim=1):
 tensor([ 0.3923,  0.2985, -0.0177])
Standard deviation (dim=1):
 tensor([0.3586, 0.6664, 1.2173])
Cumulative sum (dim=1):
 tensor([[ 0.3086,  0.3017,  1.1621,  1.5692],
        [ 0.4686,  1.6307,  1.3047,  1.1939],
        [ 1.6072,  1.8263,  0.9697, -0.0706]])


**Nhận xét (Yêu cầu Bài 7):**
- Khi dùng `dim=0`, các hàm thống kê sẽ tính toán theo chiều dọc (gom các hàng lại), tức là tính theo từng cột. Kết quả trả về gồm 4 giá trị.
- Khi dùng `dim=1`, các hàm thống kê sẽ tính toán theo chiều ngang (gom các cột lại), tức là tính trên từng hàng riêng biệt. Kết quả trả về gồm 3 giá trị.

### Bài 8. Autograd cơ bản

In [8]:
import torch

# Phần 1: y = x^2
x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()

print("--- Phần y = x^2 ---")
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

# Phần 2: YÊU CẦU: Đổi hàm thành y = 3x^2 + 2x + 1
x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * (x2 ** 2) + 2 * x2 + 1
y2.backward()

print("\n--- Phần y = 3x^2 + 2x + 1 ---")
print("x2 =", x2)
print("y2 =", y2)
print("x2.grad =", x2.grad)

--- Phần y = x^2 ---
x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)

--- Phần y = 3x^2 + 2x + 1 ---
x2 = tensor(3.6000, requires_grad=True)
y2 = tensor(47.0800, grad_fn=<AddBackward0>)
x2.grad = tensor(23.6000)


**Tính thủ công (Yêu cầu Bài 8):**
1. Hàm $y = x^2$ -> Đạo hàm $y' = 2x$. Tại $x = 3.6$ -> $y' = 2 \times 3.6 = 7.2$. Kết quả khớp với `x.grad`.
2. Hàm $y = 3x^2 + 2x + 1$ -> Đạo hàm $y' = 6x + 2$. Tại $x = 3.6$ -> $y' = 6 \times 3.6 + 2 = 23.6$. Kết quả khớp với `x2.grad`.

### Bài 9. Gradient Descent tìm hệ số đa thức

In [9]:
import numpy as np
import torch

# Standard polynomial y = x^2 + 2x + 3
polynomial = np.poly1d([1, 2, 3])

N = 20
X = np.random.randn(N, 1)
Y = polynomial(X)

XX = np.hstack((X * X, X, np.ones_like(X)))

x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

w = torch.randn(3, 1, requires_grad=True)
print("Initial coefficients:\n", w)

optimizer = torch.optim.NAdam([w], lr=0.01)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y_pred - y))
    mse.backward()
    optimizer.step()

print("\nLearned coefficients:\n", w)

Initial coefficients:
 tensor([[-0.8325],
        [ 0.0902],
        [ 0.4985]], requires_grad=True)

Learned coefficients:
 tensor([[0.8231],
        [1.4257],
        [2.8796]], requires_grad=True)


**Nhận xét (Yêu cầu Bài 9):**
- Ba hệ số học được rất xấp xỉ với hệ số gốc `[1, 2, 3]` của đa thức (sẽ có sai số cực nhỏ do bản chất xấp xỉ của Gradient Descent).
- **Vai trò của các hàm optimizer:**
  - `optimizer.zero_grad()`: Xóa sạch (reset về 0) các gradient đã được tích lũy từ vòng lặp trước. Nếu không, gradient sẽ bị cộng dồn sai lệch.
  - `mse.backward()`: Kích hoạt cơ chế Autograd tính đạo hàm ngược, tìm ra gradient của hàm loss (`mse`) theo từng trọng số (`w`).
  - `optimizer.step()`: Cập nhật giá trị mới cho các trọng số (`w`) dựa vào gradient vừa tính được và tốc độ học (`lr`) để giảm dần sai số.

### Bài 10. Giải hệ phương trình bằng Autograd

In [10]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    optimizer.zero_grad()
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    sqerr.backward()
    optimizer.step()

print(f"Nghiệm tìm được: A = {A.item():.4f}, B = {B.item():.4f}, C = {C.item():.4f}, D = {D.item():.4f}")

print("\n--- Kiểm tra lại phương trình ---")
print(f"A + B = {A.item() + B.item():.4f} (Đích: 9)")
print(f"C - D = {C.item() - D.item():.4f} (Đích: 1)")
print(f"A + C = {A.item() + C.item():.4f} (Đích: 8)")
print(f"B - D = {B.item() - D.item():.4f} (Đích: 2)")

Nghiệm tìm được: A = 4.1022, B = 4.8978, C = 3.8978, D = 2.8978

--- Kiểm tra lại phương trình ---
A + B = 8.9999 (Đích: 9)
C - D = 1.0000 (Đích: 1)
A + C = 8.0000 (Đích: 8)
B - D = 2.0000 (Đích: 2)


**Giải thích (Yêu cầu Bài 10):**
Bài toán này có **vô số nghiệm** (nhiều nghiệm) bởi vì các phương trình trong hệ phụ thuộc tuyến tính lẫn nhau. 
Lấy phương trình (1) + (2): $(A + B) + (C - D) = 9 + 1 = 10 \Rightarrow A + B + C - D = 10$
Lấy phương trình (3) + (4): $(A + C) + (B - D) = 8 + 2 = 10 \Rightarrow A + B + C - D = 10$
Thực chất chỉ có 3 phương trình độc lập nhưng lại có 4 ẩn, dẫn tới hệ có một bậc tự do (đường thẳng các nghiệm). Việc Gradient Descent dừng lại ở nghiệm nào phụ thuộc hoàn toàn vào giá trị khởi tạo ngẫu nhiên ban đầu của A, B, C, D.

---
## 3. Bài tập áp dụng
### Bài 11. Các yêu cầu tổng hợp

In [11]:
import torch
import numpy as np

print("--- 11.1 Tạo Tensor X và Reshape ---")
X = torch.arange(24) # Tạo tensor có 24 phần tử
print("Reshape 4x6:\n", X.reshape(4, 6))
print("Reshape 2x12:\n", X.reshape(2, 12))
print("Reshape 2x3x4:\n", X.reshape(2, 3, 4))

print("\n--- 11.2 Các phép toán trên A, B (3x4) ---")
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("A + B:\n", A + B)
print("A * B:\n", A * B)
print("A / B:\n", A / B)
print("Trung bình theo cột (A):\n", torch.mean(A, dim=0))
print("Độ lệch chuẩn theo cột (A):\n", torch.std(A, dim=0))

print("\n--- 11.3 Autograd cho y = 2x^2 + 5x + 3 ---")
x = torch.tensor(2.0, requires_grad=True)
y = 2 * (x**2) + 5*x + 3
y.backward()
print("PyTorch dy/dx tại x=2:", x.grad.item())
# Tính thủ công: y' = 4x + 5. Tại x = 2 -> 4(2)+5 = 13
print("Thủ công dy/dx tại x=2: 13") 

print("\n--- 11.4 Huấn luyện với y = 2x^2 - 3x + 5 ---")
poly_new = np.poly1d([2, -3, 5])
Y_new = poly_new(X_np := np.random.randn(20, 1))
XX_new = np.hstack((X_np * X_np, X_np, np.ones_like(X_np)))

x_t = torch.tensor(XX_new, dtype=torch.float32)
y_t = torch.tensor(Y_new, dtype=torch.float32)
w_t = torch.randn(3, 1, requires_grad=True)

opt = torch.optim.NAdam([w_t], lr=0.01)
for _ in range(1500):
    opt.zero_grad()
    mse = torch.mean(torch.square(x_t @ w_t - y_t))
    mse.backward()
    opt.step()

print("Hệ số tìm được (kỳ vọng [2, -3, 5]):\n", w_t.detach().numpy())

--- 11.1 Tạo Tensor X và Reshape ---
Reshape 4x6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
Reshape 2x12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
Reshape 2x3x4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

--- 11.2 Các phép toán trên A, B (3x4) ---
A + B:
 tensor([[-0.8820,  1.4735, -2.2477,  0.3115],
        [-1.3119,  0.4646,  0.3115, -1.8745],
        [-0.0890, -0.1796, -0.0561,  1.9854]])
A * B:
 tensor([[-2.3452e-01,  5.3865e-01, -3.7449e-01, -7.1809e+00],
        [-7.8527e-01, -2.3958e-01, -1.0303e-01,  7.1700e-01],
        [-1.6119e-01, -1.2033e-01, -1.0764e+00, -3.5850e-04]])
A / B:
 tensor([[-5.1219e+00,  1.1915e+00, -1.5426e+01, -1.1232e+00],
        [-3.9379e+00, -3.9985e-01, -3.9224